### PR TITLE
Demo usage of Gauge test framework integration with Browser Agent and OpenTelemetry [SLDEV-14120]

### PR DESCRIPTION
**Pull request by kristijanstefanoski**

### CODE REVIEW

File: gauge-browser-agent-otel/package.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "gauge-taiko-template",
```
Comment (yishayh) [inline gauge-browser-agent-otel/package.json]: Please update name and description.

File: gauge-browser-agent-otel/tests/step_implementation.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        Authorization: `Bearer ${config.apiToken}`,
+      },
+    }
+  );
+};
+
+// Works the same with any test framework, before test starts emit event to set baggage
+beforeScenario(async (scenario) => {
+  // Start a test session
+  const { testSessionId } = (await createTestSession()).data.data;
```
Comment (yishayh) [inline gauge-browser-agent-otel/tests/step_implementation.js]: We need to split this in the demo code according to [Gauge's life-cycle](https://github.com/getgauge/gauge-js/blob/master/docs/syntax/execution-hooks.md) and add support to all the of the TIA public API life-cycle:
1.  Creating a test-session should be done in the beforeSuite.
2. We need to add a call to the DELETE `/sl-api/v1/test-sessions/:testSessionId` at the afterSuite.
3. We need to add a call to the POST `/sl-api/v2/test-sessions/:testSessionId` in the beforeScenario for the test-start
4. We need to add a call to the POST `/sl-api/v2/test-sessions/:testSessionId` in the afterScenario for the test-end

File: gauge-browser-agent-otel/tests/step_implementation.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+          "x-sl-test-session-id": args.testSessionId,
+        },
+      });
+      window.dispatchEvent(customEvent);
+    },
+    { args: { scenario, testSessionId } }
+  );
+});
+
+beforeSuite(async () => {
```
Comment (yishayh) [inline gauge-browser-agent-otel/tests/step_implementation.js]: We need to add dispatch of a "unset:baggage" at `afterScenario`

File: gauge-browser-agent-otel/tests/step_implementation.js [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+const decoded = jwtDecode(config.apiToken); // Agent Token
+const baseUrl = decoded["x-sl-server"]; // Base url of the backend
+
+const createTestSession = () => {
+  return axios.post(
+    baseUrl.replace("/api", "/sl-api/v1/test-sessions"), // Public API to start a test session
+    {
+      testStage: "Gauge Tests",
+      bsid: "51721270-ead5-498b-b22f-c3f9861ea44e", // Better set from environment variable or get from window.$Sealights
```
Comment (yishayh) [inline gauge-browser-agent-otel/tests/step_implementation.js]: The sample code needs to have this coming from an environment variable or according to any other best-practice used by Gauga.

**Review by yishayh — CHANGES_REQUESTED** [review]
Nice job :)
Please review my comments.

File: gauge-browser-agent-otel/README.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
+# Quick start
+This is a short demo project that uses the [browser-agent](https://github.com/Sealights/SL.OnPremise.Agent.JavaScript/tree/main/browser-agent) that has integrated inside the custom [OTEL Plugin](https://github.com/Sealights/SL.OnPremise.BrowserAgent.Otel.CustomPlugin).
```
Comment (alonweiss-sl) [inline gauge-browser-agent-otel/README.md]: Please update the link to point to the official browser documentation. The github repo is private and cannot be accessed by others./

Comment (alonweiss-sl) [reply gauge-browser-agent-otel/README.md]: Same for the https://github.com/Sealights/SL.OnPremise.BrowserAgent.Otel.CustomPlugin link -- it should be removed as the contents were merged into the former repo

File: gauge-browser-agent-otel/README.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+```
+
+A sample calculator backend project also exists in this demo under the `backend` folder. The calculator
+frontend application (already built) can be found inside the `frontend` folder.
+
+**IMPORTANT** : Make sure to scan the frontend application with the latest OTEL flavored browser agent prior to running the demo!
+
+To run this demo from scratch:
+```bash
+npm i -g httpster
```
Comment (alonweiss-sl) [inline gauge-browser-agent-otel/README.md]: you can skip installation and replace line 33 with "npx httpster"

File: gauge-browser-agent-otel/tests/step_implementation.js (hunk 2 above)
Comment (kristijanstefanoski) [reply gauge-browser-agent-otel/tests/step_implementation.js]: For reporting test events I guess the API is changed to v2?
`POST /sl-api/v2/test-sessions/{testSessionId}` ?

https://sealights.atlassian.net/wiki/spaces/SUP/pages/2690220039/Test+Sessions+API+a.k.a+TIA+API#Sending-Test-Events

Comment (kristijanstefanoski) [reply gauge-browser-agent-otel/tests/step_implementation.js]: Also do we need to call this API twice? Or just once in the afterScenario like so:

`
        POST /sl-api/v2/test-sessions/{testSessionId}
        Authorization: Bearer {sealights-agent-token}
        Content-Type: application/json
        [
          {
            "name": "TestToRun1",
            "start": 1619341754996,
            "end": 1619341754998,
            "status": "passed"
          }
        ]
`

Comment (yishayh) [reply gauge-browser-agent-otel/tests/step_implementation.js]: 1. Yes, I've updated my comment to v2.
2. You can call it once, but calling it twice will allow us catching an onboarding issue easier.

File: gauge-browser-agent-otel/services/sealightsService.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+const jwtDecode = require("jwt-decode");
+const fs = require("fs");
+
+const apiToken = fs.readFileSync(`${__dirname}/../sltoken.txt`, "utf-8");
+const bsid = fs.readFileSync(`${__dirname}/../buildSessionId`, "utf-8");
+
+const decoded = jwtDecode(apiToken); // Agent Token
+const baseUrl = decoded["x-sl-server"]; // Base url of the backend
+
+const http = axios.create({
```
Comment (yishayh) [inline gauge-browser-agent-otel/services/sealightsService.js]: Better use a specific name, for xample testSessionsV1Instance  and testSessionsV2Instance

File: gauge-browser-agent-otel/tests/step_implementation.js [hunk 8]
```diff
@@ -15,64 +15,68 @@ const {
...
+    window.dispatchEvent(customEvent);
   });
-  await goto("http://localhost:3333");
+  // Send test event to Sealights
+  const data = await SLService.sendTestEvent(
+    testSession,
+    scenario.currentScenario.name,
+    testStartTime,
+    Date.now(),
+    scenario.isFailed ? "failed" : "passed" // could also be "skipped" value, implement logic as needed
```
Comment (yishayh) [inline gauge-browser-agent-otel/tests/step_implementation.js]: Can we please add the skipped as well?

File: gauge-browser-agent-otel/README.md [hunk 9]
```diff
@@ -1,35 +1,101 @@
...
+  // Unset baggage after scenario
+  await evaluate("", async () => {
+    const customEvent = new CustomEvent("delete:baggage");
+    window.dispatchEvent(customEvent);
+  });
+});
+```
+
+## 4. Sending test events
+We can send test events to Sealights in the `beforeScenario` and/or `afterScenario` hook, again using Sealights Public API like so:
```
Comment (yishayh) [inline gauge-browser-agent-otel/README.md:71]: You forgot to include an example code.

**Review by yishayh — COMMENTED** [review]
Very nice :)
small comments

**Review by yishayh — DISMISSED** [review]

File: gauge-browser-agent-otel/tests/step_implementation.js (hunk 8 above)
Comment (kristijanstefanoski) [reply gauge-browser-agent-otel/tests/step_implementation.js]: I tried to but looks like there isn't a way to skip tests in `Gauge`, at least not in their docs... so I remove this comment as it is misleading. The only way to skip a test is if some error was thrown (like step wasn't implemented) but in that case it won't even execute any of the hooks.

Still not implemented as per this: https://github.com/getgauge/gauge/issues/1397
There is no way currently to skip execution at run time.

Comment (alonweiss-sl) [reply gauge-browser-agent-otel/tests/step_implementation.js]: This is out of scope for this task and was not in the DoD.

**Comment by kristijanstefanoski** [comment]
@alonweiss-sl @yishayh Updated PR README with instructions for latest `browser-agent` version using `--enableOpenTelemetry` flag during scan

File: gauge-browser-agent-otel/package.json [hunk 10]
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "sl-browser-agent-gauge-integration",
+  "description": "Sample project demonstrating usage of Sealights Browser Agent with OTEL flavor and Gauge with taiko tests",
+  "scripts": {
+    "start": "cd calculator-app && httpster",
```
Comment (alonweiss-sl) [inline gauge-browser-agent-otel/package.json]: httpster should be a devDependency, no?

**Review by alonweiss-sl — CHANGES_REQUESTED** [review]
One final comment

File: gauge-browser-agent-otel/package.json (hunk 10 above)
Comment (kristijanstefanoski) [reply gauge-browser-agent-otel/package.json]: Yeah this script is not valid anymore, I removed it. It's explained in the README to use npx

**Comment by marcam-sl** [comment]
@alonweiss-sl Why is thisPR in the Public Repo of integration sample with TIA API?

**Review by alonweiss-sl — APPROVED** [review]

**Pull request merged**